In [288]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import train_test_split
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import sent_tokenize, word_tokenize
import re
from nltk.corpus import stopwords

#### Importation des données

In [289]:
df = pd.read_csv("C:/Users/hp/Desktop/MASTER_MSD/S2/Web_Mining/TPs/Atelier - processing and sentiment analysis/Youtube01-Psy.csv")

In [290]:
df_data = df[['CONTENT', 'CLASS']]

In [315]:
df_data.head(5)

CONTENT  CLASS
0  Huh, anyway check out this you[tube] channel: ...      1
1  Hey guys check out my new channel and our firs...      1
2             just for test I have to say murdev.com      1
3   me shaking my sexy ass on my channel enjoy ^_^       1
4            watch?v=vtaRGgvGtWQ   Check this out .      1

#### Séparer les caractéristiques et la cible

In [292]:
df_x = df_data['CONTENT']
df_y = df_data.CLASS

In [293]:
#Convertir df_x_content en une liste
df_x_content=df_x.tolist()

### Question1:Ajoutez d’autres prétraitement afin d’améliorer le modèle

### Preprocessing

In [294]:
processed_data_list=[]
for data in df_x_content :
    processed_data = data.lower()
    processed_data = re.sub('[^a-zA-Z]', ' ', processed_data )
    processed_data = re.sub(r'\s+', ' ', processed_data)
    processed_data_list.append(processed_data)

all_data_words=[]
lemmmatizer=WordNetLemmatizer()
for data in processed_data_list:
    data1=data
    words = word_tokenize(data1)
    words = [lemmmatizer.lemmatize(word.lower()) for word in words if(not word in set(stopwords.words('english')) 
                                                                      and  word.isalpha())]
    all_data_words.append(words)
 

#### Feature Extracion

In [295]:
# Extract the features with countVectorizer
all_data_words=pd.Series(all_data_words).astype(str)
cv = CountVectorizer()
X = cv.fit_transform(all_data_words)

#### Train test split

In [296]:
X_train, X_test, y_train, y_test = train_test_split(X, df_y, test_size=0.33, random_state=42)

#### Modèle Naive Bayes

In [297]:
clf = MultinomialNB()
clf.fit(X_train, y_train)

MultinomialNB()

#### Tester le modèle

In [298]:
clf.score(X_test, y_test)

0.9482758620689655

### Question2:Refaites le même travail en utilisant TFIDF vectorizer avec la régression logistique 

In [299]:
# Feature extraction avec TFIDF
from sklearn.feature_extraction.text import TfidfVectorizer
all_data_words=pd.Series(all_data_words).astype(str)
tfidf_vectorizer = TfidfVectorizer()
tfidf_matrix = tfidf_vectorizer.fit_transform(all_data_words)

scipy.sparse.csr.csr_matrix

In [300]:
X_train, X_test, y_train, y_test = train_test_split(tfidf_matrix, df_y, test_size=0.33, random_state=42)
type(X_train)

scipy.sparse.csr.csr_matrix

In [301]:
from sklearn.linear_model import LogisticRegression

#### Resgression logistique

In [302]:
model = LogisticRegression()
model.fit(X_train, y_train)

LogisticRegression()

#### Tester le modèle

In [303]:
model.score(X_test, y_test)

0.9655172413793104

### Question3: Comparez les deux scores 

##### On remarque qu'après l'utilisation de TFIDF vectorizer avec la régression logistique , le scour est devenu plus élevé:0.965 au lieu de 0.948  


### Question4: Utilisation d'un nouveau source de données

#### a- Faites le prétraitement en utilisant word2vec avec l’implémentation skip-gram

In [305]:
# Importation des données
df = pd.read_csv("C:/Users/hp/Desktop/MASTER_MSD/S2/Web_Mining/TPs_Realisé/Atelier3/PDF_Flask/train_E6oV3lV.csv")

In [306]:
df.head(5)

id  label                                              tweet
0   1      0   @user when a father is dysfunctional and is s...
1   2      0  @user @user thanks for #lyft credit i can't us...
2   3      0                                bihday your majesty
3   4      0  #model   i love u take with u all the time in ...
4   5      0             factsguide: society now    #motivation

In [307]:
df.shape

(31962, 3)

In [309]:
df = df[:10000] # Réduire la taille

In [310]:
import numpy as np
def remove_pattern(input_txt, pattern):
    r = re.findall(pattern, input_txt)
    for i in r:
        input_txt = re.sub(i, '', input_txt)
    return input_txt

In [311]:
#Removing Twitter Handles (@user)
df['tweet'] = np.vectorize(remove_pattern)(df['tweet'], "@[\w]*") 
df.tweet.head(10)

0      when a father is dysfunctional and is so sel...
1      thanks for #lyft credit i can't use cause th...
2                                  bihday your majesty
3    #model   i love u take with u all the time in ...
4               factsguide: society now    #motivation
5    [2/2] huge fan fare and big talking before the...
6                     camping tomorrow        danny…
7    the next school year is the year for exams.ðŸ˜...
8    we won!!! love the land!!! #allin #cavs #champ...
9                welcome here !  i'm   it's so #gr8 ! 
Name: tweet, dtype: object

In [312]:
# Features and Labels
df_x = df['tweet']
df_y = df.label

#### Prétraitement

In [313]:
#Convertir df_x en liste
df_x_content=df_x.tolist()

In [314]:
processed_data_list=[]
for data in df_x_content :
    processed_data = data.lower()
    processed_data = re.sub('[^a-zA-Z]', ' ', processed_data )
    processed_data = re.sub(r'\s+', ' ', processed_data)
    processed_data_list.append(processed_data)

all_data_words=[]
lemmmatizer=WordNetLemmatizer()
for data in processed_data_list:
    data1=data
    words = word_tokenize(data1)
    words = [lemmmatizer.lemmatize(word.lower()) for word in words if(not word in set(stopwords.words('english'))
                                                                      and  word.isalpha())]
    all_data_words.append(words) 

#### Extraction des caractéristiques

In [271]:
# Extract the features with countVectorizer
all_data_words=pd.Series(all_data_words).astype(str)
cv = CountVectorizer()
X = cv.fit_transform(all_data_words)

#### Train test split

In [272]:
X_train, X_test, y_train, y_test = train_test_split(X, df_y, test_size=0.33, random_state=42)

#### b- Entrainer le modèle en utilisant  Regression Logistique

In [273]:
model = LogisticRegression()
model.fit(X_train, y_train)

LogisticRegression()

In [274]:
model.score(X_test, y_test)

0.943939393939394

In [275]:
clf = MultinomialNB()
clf.fit(X_train, y_train)
clf.score(X_test, y_test)

0.933030303030303